_posts/2018-03-30-angular-security-best-practices.md

```diff
@@ -0,0 +1,151 @@
+Otherwise, the value will be sanitized to be safe according to the security context.
+
+There are three main helper functions for sanitizing the values.
+The `sanitizeHtml` function sanitizes the untrusted HTML value by parsing the value and checks its tokens.
+The `sanitizeStyle` and `sanitizeUrl` functions sanitize the untrusted style or URL value by regular expressions.
+
+#### How can we disable the sanitization logic?
+In specific situations, it might be necessary to disable sanitization.
+Users can bypass security by constructing a value with one of the `bypassSecurityTrustX` methods, and then binding to that value from the template.
+
+> Be careful: If you trust a value that might be malicious, you are introducing a security vulnerability into your application!
+
+
+## 3. Content security policy (CSP)
+
+Content Security Policy is a defense-in-depth technique to prevent XSS.
+To enable CSP, configure your web server to return an appropriate `Content-Security-Policy` HTTP header.
+More info about specifying your CSP can be found on [Google Developers Page about Content Security Policy](https://developers.google.com/web/fundamentals/security/csp/){:target="blank" rel="noopener noreferrer"}.
+To check if your CSP is valid you can use the [CSP evaluator from google](https://csp-evaluator.withgoogle.com){:target="blank" rel="noopener noreferrer"}.
+
+
+## 4. Use the offline template compiler
+
+_Angular templates_ are the same as executable code: HTML, attributes, and binding expressions (but not the values bound) in templates are trusted to be safe.
+This means that if an attacker can control a value that is being parsed by the template we have a security leak.
+Never generate template source code by concatenating user input and templates.
+To prevent these vulnerabilities, use the **offline template compiler**, also known as _template injection_.
+
+
+## 5. Avoid direct use of the DOM APIs
+
+The built-in browser DOM APIs don't automatically protect you from security vulnerabilities.
+For example, _document_, the node available through _ElementRef_, and many _third-party APIs_ contain unsafe methods.
+Avoid interacting with the DOM directly and instead use **Angular templates** where possible.
+
+
+## 6. Server-side XSS protection
+
+Injecting template code into an Angular application is the same as injecting executable code into the application.
+So, validate all data on server-side code and escape appropriately to prevent XSS vulnerabilities on the server.
+Also, Angular recommends _not to generate Angular templates on the server side_ using a templating language.
+
+
+# HTTP-level vulnerabilities
+
+Angular has built-in support to help prevent two common HTTP vulnerabilities, cross-site request forgery (CSRF or XSRF) and cross-site script inclusion (XSSI).
+Both of these must be mitigated primarily on the server side, but Angular provides helpers to make integration on the client side easier.
```

Comment:
How do they do that?

Reply:
That is explained in the next 2 chapters